File: tools/terrain-crystal-tests/lava.py
"""CPU execution of production UV addressing, not a GPU render test."""
from pathlib import Path
import re
import subprocess
import tempfile
root=Path(__file__).resolve().parents[2]
shim=(root/'tools/lighting-tests/NativeTransportShim.cpp').read_text()
extra='''
float clamp(float a,float b,float c){return std::clamp(a,b,c);}
float2 clamp(float2 a,float2 b,float2 c){return {clamp(a.x,b.x,c.x),clamp(a.y,b.y,c.y)};}
float2 clamp(float2 a,float b,float2 c){return clamp(a,make_float2(b),c);}
float2 max(float2 a,float b){return max(a,make_float2(b));}
'''
source='\n'.join((root/'Assets/Shaders'/n).read_text() for n in ['TerrainTileAddressing.hlsl','TerrainSampling.hlsl'])
source=re.sub(r'^#.*$', '',source,flags=re.M)
source=re.sub(r'\b(float[234])\(',r'make_\1(',source)
# Scalar vector splats are implicit in HLSL, explicit with clang vectors.
source=re.sub(r'(res\.(?:tileOffsetUV|availableTileSize|finalUV|minTileUV|maxTileUV)) = 0.0;',r'\1 = make_float2(0.0);',source)
with tempfile.TemporaryDirectory(prefix='kern-lava-') as tmp:
    cpp=Path(tmp)/'test.cpp';exe=Path(tmp)/'test'
    for mutation in (False,True):
        candidate=source.replace('if ((int)(animData.w + 0.5) == 2)','if (false)') if mutation else source
        cpp.write_text(shim+extra+candidate+(Path(__file__).parent/'lava.cpp').read_text())
        subprocess.run(['clang++','-std=c++20','-O2',str(cpp),'-o',str(exe)],check=True)
        result=subprocess.run([str(exe)])
        assert result.returncode != 0 if mutation else result.returncode == 0
print('Old cell-local lava addressing mutation rejected.')

File: tools/terrain-crystal-tests/lava.cpp
float2 lava(float2 cell, float2 local, float time, float anchored, float descriptor) {
    auto tile=ResolveTerrainTileUV(
        make_float2(.7f,.3f), make_float4(.25f,.125f,.15625f,.125f),
        make_float4(.015625f,.015625f,1,8), make_float4(cell.x,cell.y,descriptor,1),
        make_float4(4,10,0,2), make_float4(anchored,local.x,local.y,0), time,
        make_float2(1.f/2048));
    return ClampTerrainTileUV(tile.finalUV,tile);
}
int main() {
    int checks=0;
    for(int x : {-32,-1,0,9,31,32})
    for(int y : {-32,-1,0,7,31,32})
    for(float time : {0.f,.17f,1.7f,19.3f})
    for(int i=0;i<32;i++) {
        float t=(i+.5f)/32;
        auto a=lava(make_float2(x,y),make_float2(1.125f,t),time,1,3);
        auto b=lava(make_float2(x+1,y),make_float2(.125f,t),time,0,7);
        auto c=lava(make_float2(x,y),make_float2(t,-.125f),time,1,3);
        auto d=lava(make_float2(x,y+1),make_float2(t,.875f),time,0,7);
        if(dot(a-b,a-b)>1e-10 || dot(c-d,c-d)>1e-10) return 1;
        checks++;
    }
    auto a=lava(make_float2(2,3),make_float2(.5f,.5f),0,0,0);
    auto b=lava(make_float2(2,3),make_float2(.5f,.5f),1,0,0);
    if(dot(a-b,a-b)<1e-6) return 2;
    printf("Lava production UV math: %d adjacency checks, carrier/descriptor independence and motion passed.\n", checks);
}
